storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+package storage
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/containers/image/image"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/archive"
+	"github.com/containers/storage/pkg/ioutils"
+	"github.com/containers/storage/storage"
+	ddigest "github.com/docker/distribution/digest"
+)
+
+var (
+	// ErrInvalidBlobDigest is returned when PutBlob() is given a blob
+	// with a digest-based name that can't be used as a map key.
+	ErrInvalidBlobDigest = errors.New("invalid blob digest")
+	// ErrBlobDigestMismatch is returned when PutBlob() is given a blob
+	// with a digest-based name that doesn't match its contents.
+	ErrBlobDigestMismatch = errors.New("blob digest mismatch")
+	// ErrManifestDigestMismatch is returned when GetTargetManifest() is
+	// given a digest-based name that doesn't match the manifest.
+	ErrManifestDigestMismatch = errors.New("manifest digest mismatch")
+	// ErrNoSuchImage is returned when we attempt to access an image which
+	// doesn't exist in the storage area.
+	ErrNoSuchImage = storage.ErrNotAnImage
+)
+
+type storageImage struct {
```

Comment:
Is there a benefit to sharing a single object for the `ImageSource` and `ImageDestination` implementation?

It kinda allows a reader to see the in-progress contents of an image like `BlobData` before `ImageDestination.Commit` I guess, OTOH that would only be beneficial in `storageImage`-specific code; generic callers would call `storageReference.NewImageSource` and `storageReference.NewImageDestination` separately, and get two separate `storageImage` objects, referring to the same `reference` or `id`, but without a shared pending state.